### ISSUE TITLE
Using HTTPS through proxy break asset URLs and CP AJAX requests

### ISSUE THREAD

**Issue by FrittenKeeZ** [statamic/v2-hub#1586]
### Expected behaviour
Having a site running with HTTPS through proxy (fx. a load balancer forwarding traffic from 443 to 80), all links and assets with absolute URLs should work correctly, also in Control Panel.

### Actual behaviour
Assets URLs `{{ permalink }}` and CP AJAX requests are rendered without HTTPS.

### Steps to reproduce
1. Have a site running HTTPS through proxy
2. Specify correct site URL in settings
3. See that asset `{{ permalink }}` is rendered without HTTPS
4. Login to CP and see that content for a page doesn't get loaded - failed request

### Server Details

**Operating System**: Alpine Linux 3.6.2

**Web Server:** nginx 1.12.1

**PHP Version:** 7.1.5

**Statamic Version:** 2.6.8

**Updated from an older Statamic or fresh install:** 2.6.6

### Bonus
To fix the issue for CP AJAX requests, add `use_https: true` to `site/settings/system.yaml`.
To fix the issue for assets, add `HTTPS=on` to `.env` and ensure that your server passes variables on to `$_SERVER`,

The system should check if `$_SERVER['HTTP_X_FORWARDED_PROTO']` is set to `https` and adjust URLs accordingly.

**Comment by jasonvarga** [comment]
We should add support for the trusted proxies package.
https://github.com/fideloper/TrustedProxy

**Comment by sheldonkotyk** [comment]
I'm not getting any failed requests in my logs but with use_https: true, the CP still has issues pulling the submenus. For example, the settings link opens the system page fine but the submenu doesn't show the other settings options.

Same issue with collections, etc.

**Comment by FrittenKeeZ** [comment]
Any updates on this?

**Comment by nevoxx** [comment]
Do you have any news on this? We are also experiencing the issue, that submenu links are not rendered in the control panel at Statamic 2.9.6

**Comment by jasonvarga** [comment]
This issue should have been closed in 2.9.0.

In your system.yaml's locales array, does your url contain https? If not, it should.

**Comment by nevoxx** [comment]
You are right, we currently had our system.yaml looking like this, using a full url with https like `https://example.text/` fixed the issue.

```
locales:
  de:
    name: Deutsch
    full: de_DE
    url: /
```

**Comment by jasonvarga** [comment]
Great!

In 2.9, we force any generated URLs to use the base URL you've defined in there.
If you have only defined a relative one (eg `/`) then it would have to use the guessing/request-based logic, which is what can have a problem with load balancers.

For load balancers or other complicated setups, just be sure to specify your URL.

**Comment by ymakhloufi** [comment]
If you don't want to specify a FQDN in the systems.yaml (like in my case) then I recommend to add this line on top of /statamic/core/Http/routes-cp.php:

Symfony\Component\HttpFoundation\Request::setTrustedProxies(['10.0.0.0/8', '172.16.0.0/12', '192.168.0.0/16', '127.0.0.0/8', '169.254.0.0/16', 'fe80::/10', 'fc00::/7']);

If you happen to know the CIDR block of your load balancer's IP, you can even remove all other CIDR blocks from the list above.

This basically removes the need for using the fideloper/TrustedProxy package and tells Laravel (or rather Symfony) to first check the IP-address of the requesting client. If it is within the set of trusted proxies (see list above) then it checks if there is an X-Forwarded-Proto header. The load balancer sets this header to "https" when it terminates the SSL connection and forwardes the request via HTTP (non-ssl). 

TL;DR: This fix makes sure that Laravel knows that it should create routes as HTTPS, even if the request kinda looks like it looks for HTTP. And this is independent from the domain, so you don't have to hard-code the domain-name anywhere.
